PS-376 Implement WatchForChanges method

lindseysimple opened the pull request:
Implement the WatchForChanges method.

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/go-mod-configuration/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-configuration/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions
<!-- How can the reviewers test your change? -->

## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

lindseysimple committed:
PS-376 Implement WatchForChanges method

Implement the WatchForChanges method.

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

lindseysimple committed:
PS-376 Update the error handling

Update the error handling in WatchForChanges.

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

lindseysimple committed:
PS-376 Implement StopWatching method

Implement StopWatching method and add UT.

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

lindseysimple committed:
PS-376 Refactor WatchForChanges and remove keeper client wg

Refactor WatchForChanges and remove keeper client waitGroup.

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

cloudxxx8 commented on internal/pkg/keeper/client.go:
```diff
@@ -140,9 +143,80 @@ func (client *keeperClient) GetConfiguration(configStruct interface{}) (interfac
+			Messages: messages,
+		},
+	}
+	msgBusConfig := config.(*models.ConfigurationStruct).MessageQueue
```
always check error for type conversion

lindseysimple replied:
Done

cloudxxx8 commented on internal/pkg/keeper/client.go:
```diff
@@ -140,9 +143,80 @@ func (client *keeperClient) GetConfiguration(configStruct interface{}) (interfac
+			Protocol: msgBusConfig.Protocol,
+		},
+		Type:     msgBusConfig.Type,
+		Optional: msgBusConfig.Optional,
```
add a deepCopy func to produce a copy of messageBusConfig, and add a random number suffix in the MqttClientId option.
reference:
https://github.com/IOTechSystems/edgex-go-private/pull/314/files
https://github.com/IOTechSystems/edgex-go-private/blob/b95546cea1628225578aae0059b0751de3f401c8/internal/support/sparkplug/mqtt/client.go#L249-L252

lindseysimple replied:
Random number suffix in the MqttClientId added, thanks.

lindseysimple committed:
PS-376 Assign unique mqtt client before subscription

- Assign unique mqtt client before subscription
- Send a reply when watchForChanges connected initially

Signed-off-by: Lindsey Cheng <beckysocute@gmail.com>

cloudxxx8 reviewed (approved):
LGTM
